accepted/cranelift-assembler.md

```diff
@@ -0,0 +1,368 @@
+Encoding" table but we will refer to as a __format__ table:
+
+![Format (Encoding) Table](./cranelift-assembler-format.png)
+
+Our DSL merges those tables into single lines:
+
+```rust
+inst("andb", fmt("I", [rw(al), r(imm8)]), rex(0x24).ib(), None),
+inst("andw", fmt("I", [rw(ax), r(imm16)]), rex(0x25).prefix(_66).iw(), None),
+inst("andl", fmt("I", [rw(eax), r(imm32)]), rex(0x25).id(), None),
+inst("andq", fmt("I_SXLQ", [rw(rax), sxq(imm32)]), rex(0x25).w().id(), None),
+...
+inst("andb", fmt("RM", [rw(r8), r(rm8)]), rex(0x22).r(), None),
+inst("andb", fmt("RM_SXBQ", [rw(r8), r(rm8)]), rex(0x22).w().r(), None),
+inst("andw", fmt("RM", [rw(r16), r(rm16)]), rex(0x23).prefix(_66).r(), None),
+inst("andl", fmt("RM", [rw(r32), r(rm32)]), rex(0x23).r(), None),
+inst("andq", fmt("RM", [rw(r64), r(rm64)]), rex(0x23).w().r(), None),
+```
+
+Some explanation:
+- an `inst` is built up from an __instruction mnemonic__, a __format__ (`fmt`), an __encoding__
+  (`rex` here), and CPUID __features__ (`None` here)
+- the __format__ describes the instruction operands precisely and has a __format name__
+- the __encoding__ describes how to emit the machine code and eventually would have `vex` and `evex`
+  constructors
+- the __feature__ set will eventually allow one to write expressions like `AVX512_X | AVX512_Y`
+
+Note that this DSL will demand that instruction names be unique: but we used `andb` twice above! To
+solve this, the DSL concatenates the __instruction mnemonic__ and the __format name__. E.g., we
+would refer to the first instruction above as `Inst::andb_i`. Unfortunately, this plan is almost,
+but not quite, sufficient for uniqueness; we manually append descriptors like `SXBQ` where the
+table-provided names are simply not specific enough. Perhaps we could auto-append this suffix in the
+future, but we want it to be quite clear what instruction name will be generated in the next step.
+
+### Generation
+[Generation]: #generation
+
+Using the instructions defined in the DSL, we generate Rust and ISLE code. This generated code is
+able to:
+- encode an instruction
+- display an instruction
+- check CPUID features (TODO)
+- interface with register allocation
+- interface with ISLE lowering (TODO)
+
+For example, x64 has an `andl` operation that modifies a 32-bit register-or-memory operand by
+AND-ing it with a sign-extended 8-bit immediate. Using the manual's format table, we append the `mi`
+suffix; then we additionally append `sxbl` to communicate the sign-extension. The definition looks
+like this:
+
+```rust
+inst("andl", fmt("MI_SXBL", [rw(rm32), sxl(imm8)]), rex(0x83).digit(4).ib(), None),
+```
+
+When we generate Rust code from the DSL, the assembler creates the following:
+
+```rust
+/// `andl: MI_SXBL(rm32[rw], imm8[sxl]) => 0x83 /4 ib`
+#[derive(arbitrary::Arbitrary, Clone, Debug)]
+pub struct andl_mi_sxbl {
+    pub rm32: GprMem,
+    pub imm8: Imm8,
+}
+impl andl_mi_sxbl {
```

Comment:
I think it would be useful for these to be trait methods that way generic code isn't required to use the full `enum Inst` along with the probable code bloat (from having a match arm for every instruction even if we're only using one or a few) that implies.